bizarre error message for one of the universal constructor syntaxes

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#![allow(dead_code)]
#![allow(unused_variables)]
enum Enum<'l, F> {
    Tuple(&'l F),
    Struct { f: &'l F },
}
fn main(){
    let v = 42u32;
    let t = Enum::<u32>::Struct { f: &v };
}
```

I expected to see this happen: it compiles.

Instead, this happened (on the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8410c281e538e90c9a3ba32bb4b58df0), with stable and with nightly):

```
error[E0109]: lifetime arguments are not allowed on variant `Struct`
 --> src/main.rs:9:26
  |
9 |     let t = Enum::<u32>::Struct { f: &v };
  |                          ^^^^^^
  |                          |
  |                          lifetime argument not allowed
  |                          not allowed on variant `Struct`
  |
  = note: enum variants can't have type parameters

error[E0107]: this enum takes 1 lifetime argument but 0 lifetime arguments were supplied
 --> src/main.rs:9:13
  |
9 |     let t = Enum::<u32>::Struct { f: &v };
  |             ^^^^ expected 1 lifetime argument
  |
note: enum defined here, with 1 lifetime parameter: `'l`
 --> src/main.rs:3:6
  |
3 | enum Enum<'l, F> {
  |      ^^^^ --
help: add missing lifetime argument
  |
9 |     let t = Enum::<'l, u32>::Struct { f: &v };
  |                    +++
```

It's not 100% clear to me that this syntax, with the generics applied to the enum type rather than the variant, is intended to be accepted.  I think it probably ought to be, though.  On this point I found #69363/#69356.  The Reference doesn't seem to discuss it.

Also, the error messages are nonsensical, and the suggestion doesn't compile.

Variations:
```
    let t = Enum::<'l, u32>::Tuple { 0: &v }; // use of undeclared lifetime name `'l`  [obviously]
    let t = Enum::<u32>::Tuple { 0: &v }; //  lifetime arguments are not allowed on variant `Tuple`
    let t = Enum::<'_, u32>::Tuple { 0: &v }; // //  lifetime arguments are not allowed on variant `Tuple`
    let t = Enum::<u32>::Tuple(&v);; // compiles, but of course you can't use this for a named-fields variant
    let t = Enum::Struct::<u32> { f: &v }; // compiles
    let t = Enum::Tuple::<u32> { 0: &v }; // compiles
```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs
+++ b/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs
@@ -1691,7 +1691,9 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
         // which requires object lifetime defaults.
         let type_def_id = match res {
             Res::Def(DefKind::AssocTy, def_id) if depth == 1 => Some(self.tcx.parent(def_id)),
-            Res::Def(DefKind::Variant, def_id) if depth == 0 => Some(self.tcx.parent(def_id)),
+            Res::Def(DefKind::Variant, def_id) if depth == 0 || depth == 1 => {
+                Some(self.tcx.parent(def_id))
+            }
             Res::Def(
                 DefKind::Struct
                 | DefKind::Union

--- a/compiler/rustc_resolve/src/late.rs
+++ b/compiler/rustc_resolve/src/late.rs
@@ -2158,7 +2158,15 @@ impl<'a, 'ast, 'ra, 'tcx> LateResolutionVisitor<'a, 'ast, 'ra, 'tcx> {
                 Res::Def(DefKind::AssocTy, def_id) if i + 2 == proj_start => {
                     self.r.tcx.parent(def_id)
                 }
-                Res::Def(DefKind::Variant, def_id) if i + 1 == proj_start => {
+                Res::Def(DefKind::Variant, def_id)
+                    if i + 2 == proj_start && segment.has_generic_args =>
+                {
+                    self.r.tcx.parent(def_id)
+                }
+                Res::Def(DefKind::Variant, def_id)
+                    if i + 1 == proj_start
+                        && !i.checked_sub(1).is_some_and(|i| path[i].has_generic_args) =>
+                {
                     self.r.tcx.parent(def_id)
                 }
                 Res::Def(DefKind::Struct, def_id)

--- a/compiler/rustc_resolve/src/lib.rs
+++ b/compiler/rustc_resolve/src/lib.rs
@@ -353,8 +353,7 @@ enum VisResolutionError<'a> {
 struct Segment {
     ident: Ident,
     id: Option<NodeId>,
-    /// Signals whether this `PathSegment` has generic arguments. Used to avoid providing
-    /// nonsensical suggestions.
+    /// Signals whether this `PathSegment` has generic arguments.
     has_generic_args: bool,
     /// Signals whether this `PathSegment` has lifetime arguments.
     has_lifetime_args: bool,

--- a/tests/ui/lifetimes/no-lt-elision-in-variant-path-seg-as-enum-path-seg-has-gen-args.rs
+++ b/tests/ui/lifetimes/no-lt-elision-in-variant-path-seg-as-enum-path-seg-has-gen-args.rs
@@ -1,0 +1,37 @@
+// Given an enum `Enum` with lifetime parameters, we used to lower `Enum::<…>::Variant {}` to
+// `Enum::<…>::Variant::<'_, …> {}`, i.e., synthesize `'_` lifetime args for enum variants
+// even if generic args were already passed to the enum itself.
+// Obviously, these conflicting generic arg lists were then rejected later on (by HIR ty lowering).
+//
+// Now we no longer do. Why do we synthesize these lifetimes for struct variant paths in the first
+// place while we don't do it for unit & tuple variants? Well, we later HIR-ty-lower the former in
+// "type mode" while we lower the latter in "value mode". In "type mode", we won't infer
+// elided lifetimes if the generic arg lists contains non-lifetime args which is undesirable here.
+//
+// issue: <https://github.com/rust-lang/rust/issues/108224>
+//@ check-pass
+
+fn scope<'any>() {
+    enum Ty0<'a> {
+        Unit,
+        Tuple,
+        Struct {},
+
+        Carry(&'a ()),
+    }
+
+    enum Ty1<'a, T: ?Sized> {
+        Variant,
+
+        Carry(&'a T),
+    }
+
+    let _ = Ty0::<'_>::Unit {};
+    let _ = Ty0::<'static>::Tuple {};
+    let _ = Ty0::<'any>::Struct {};
+
+    let _ = Ty1::</*'_, */()>::Variant {};
+    let _ = Ty1::<'any, ()>::Variant {};
+}
+
+fn main() {}

--- a/tests/ui/object-lifetime/object-lifetime-default-enum-variant.rs
+++ b/tests/ui/object-lifetime/object-lifetime-default-enum-variant.rs
@@ -1,0 +1,24 @@
+//@ check-pass
+
+enum Enum<'a, T: 'a + AbideBy<'a> + ?Sized> {
+    Variant,
+
+    Carry(&'a T),
+}
+
+fn scope<'any>() {
+    // We elaborate `dyn Trait` to `dyn Trait + 'any` due to bound `'a` on `T`.
+    let _ = Enum::Variant::<'any, dyn Trait> {};
+
+    // Similarly here.
+    let _ = Enum::<'any, dyn Trait>::Variant {};
+}
+
+trait Trait {}
+
+// We use this to test that a given trait object lifetime bound is
+// *exactly equal* to a given lifetime (not longer, not shorter).
+trait AbideBy<'a> {}
+impl<'a> AbideBy<'a> for dyn Trait + 'a {}
+
+fn main() {}

